extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/kernel/controller/ControllerUtils.java

```diff
@@ -0,0 +1,334 @@
+      final StageId stageId = stageDef.getId();
+      retVal.computeIfAbsent(stageId, ignored -> new TreeSet<>());
+
+      for (final int inputStageNumber : queryDefinition.getStageDefinition(stageId).getInputStageNumbers()) {
+        final StageId inputStageId = new StageId(queryDefinition.getQueryId(), inputStageNumber);
+        retVal.computeIfAbsent(stageId, ignored -> new TreeSet<>()).add(inputStageId);
+      }
+    }
+
+    return retVal;
+  }
+
+  /**
+   * Returns a mapping of stage -> stages that depend on that stage. Uses TreeMaps and TreeSets so we have a consistent
+   * order for analyzing and running stages.
+   */
+  public static Map<StageId, Set<StageId>> computeStageOutflowMap(final QueryDefinition queryDefinition)
+  {
+    final Map<StageId, Set<StageId>> retVal = new TreeMap<>();
+
+    for (final StageDefinition stageDef : queryDefinition.getStageDefinitions()) {
+      final StageId stageId = stageDef.getId();
+      retVal.computeIfAbsent(stageId, ignored -> new TreeSet<>());
+
+      for (final int inputStageNumber : queryDefinition.getStageDefinition(stageId).getInputStageNumbers()) {
+        final StageId inputStageId = new StageId(queryDefinition.getQueryId(), inputStageNumber);
+        retVal.computeIfAbsent(inputStageId, ignored -> new TreeSet<>()).add(stageId);
+      }
+    }
+
+    return retVal;
+  }
+
+  /**
+   * Whether output of a stage can possibly use {@link OutputChannelMode#MEMORY}. Returning true does not guarantee
+   * that the stage *will* use {@link OutputChannelMode#MEMORY}. Additional requirements are checked in
+   * {@link #computeStageGroups(QueryDefinition, ControllerQueryKernelConfig)}.
+   */
+  public static boolean canStreamOutput(
+      final QueryDefinition queryDefinition,
+      final int stageNumber,
+      final ControllerQueryKernelConfig config,
+      final Map<StageId, Set<StageId>> outflowMap
+  )
+  {
+    if (config.isFaultTolerant()) {
+      // Cannot stream if fault tolerance is enabled: durable storage is required.
+      return false;
+    }
+
+    if (!config.isPipeline() || config.getMaxConcurrentStages() < 2) {
+      // Cannot stream if pipelining (& running multiple stages at once) is not enabled.
+      return false;
+    }
+
+    final StageId stageId = queryDefinition.getStageDefinition(stageNumber).getId();
+    final Set<StageId> outflowStageIds = outflowMap.get(stageId);
+
+    if (outflowStageIds.isEmpty()) {
+      return true;
+    } else if (outflowStageIds.size() == 1) {
+      final StageDefinition outflowStageDef =
+          queryDefinition.getStageDefinition(Iterables.getOnlyElement(outflowStageIds));
+
+      // Two things happening here:
+      //   1) Stages cannot both stream and broadcast their output. This is because when streaming, we can only
+      //      support a single reader.
+      //   2) Stages can only receive a single streamed input. This isn't strictly necessary, but it simplifies the
```

Comment:
Ah, I really meant this to refer to output channel mode MEMORY. I have tried to make it more clear and have removed the word "stream". This comment now reads:

```
      // Two things happening here:
      //   1) Stages cannot use output mode MEMORY when broadcasting. This is because when using output mode MEMORY, we
      //      can only support a single reader.
      //   2) Downstream stages can only have a single input stage with output mode MEMORY. This isn't strictly
      //      necessary, but it simplifies the logic around concurrently launching stages.
```

I have also renamed various variables, like `canStream` to `canUseMemoryOutput`.